### PR TITLE
feat(batch): support spilling for the batch sort executor

### PR DESCRIPTION
**Pull request by chenzl25**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

- Tracking issue: https://github.com/risingwavelabs/risingwave/issues/16615
- This PR is implemented based on https://github.com/risingwavelabs/rfcs/pull/92
image

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Review by fuyufjh — APPROVED** [review]

File: src/batch/src/executor/order_by.rs
[the remark this answers is not in the document]
Comment (chenzl25) [reply src/batch/src/executor/order_by.rs:1]: I think we can give a very small memory limit and use opendal's memory backend to test these code paths later.

File: src/batch/src/executor/order_by.rs [hunk 1]
```diff
@@ -71,76 +88,199 @@ impl BoxedExecutorBuilder for SortExecutor {
...
-                .sum::<usize>();
-            encoded_rows.extend(
-                encoded_chunk
-                    .into_iter()
-                    .enumerate()
-                    .map(|(row_id, row)| (chunk.row_at_unchecked_vis(row_id), row)),
-            );
-            if !self.mem_context.add(chunk_estimated_heap_size as i64) {
-                Err(BatchError::OutOfMemory(self.mem_context.mem_limit()))?;
+        if !need_to_spill {
```
[the remark this answers is not in the document]
Comment (chenzl25) [reply src/batch/src/executor/order_by.rs:139]: Sure. I can remove the first bracket `if !need_to_spill`, it won't affect the correctness.

File: src/batch/src/executor/order_by.rs
[the remark this answers is not in the document]
Comment (fuyufjh) [reply src/batch/src/executor/order_by.rs:1]: Even with a very small memory limit, we might still need to generate thousands of rows to trigger a disk spill, which can't be met in most E2E cases. I am thinking is there any other ways for it? For example, introducing a config or session parameter to ensure spill-out is used regardless of the data volume, and then run some query.

[the remark this answers is not in the document]
Comment (chenzl25) [reply src/batch/src/executor/order_by.rs:1]: The very small memory limit could be zero actually.

**Pull request merged**